Database diagram

kinow opened the pull request:
Closes #28 

Adds script `bin/database-diagram.py` which relies on `eralchemy`, to produce a graphviz/dot database diagram.

The script uses `cylc-flow` Python code to create the SQLite DB, just like what Cylc does normally. Then `eralchemy` produces a dot file.

The diagram generated by `eralchemy` is thin and long, without any constraints. Adding the contraints to `cylc-flow`'s DB could lead to regressions when running suites, so I prefer to leave that for post Cylc 8 - or at least until I feel confident to help in case things go bad after this change. `eralchemy` supports manually entered relationships, so they are hardcoded for now (we don't seem to change the DB schema with frequency anyway).

To solve the confusing layout in the vanilla diagram, the script re-organizes the contents of the `eralchemy` generated dot file, by changing the graph settings, and adding a subgraph.

Final diagram:

![cylc-database](https://user-images.githubusercontent.com/304786/64055225-4b1f5180-cbdf-11e9-90cc-d25c9b76797b.png)


Added a section to the current documentation as well. Later when we start working on the new documentation, I would like to move the database documentation up to the initial chapters, as I believe this is basic/introductory information, which is currently listed under "Suite Run databases", in chapter 13.

![image](https://user-images.githubusercontent.com/304786/64055574-668b5c00-cbe1-11e9-8dda-e9f0ef000aa5.png)


ps: `eralchemy` is not a hard requirement for the documentation, but I've added an optional dependency under group `[database]`, that can be used only when the script needs to be executed again via `pip install .[database]` and then `python bin/database-diagram.py`.

matthewrmshin commented:
Looks good.

kinow commented:
Travis failed with  `ModuleNotFoundError: No module named 'sphinx'`, but I think it is not due to changes in this PR. It built correctly in my notebook

kinow commented:
Conflicts fixed, updated `setup.py`'s commit to match new syntax from `master`. Rebased, ready for review again.

sadielbartholomew commented on bin/database-diagram.py:
```diff
@@ -0,0 +1,236 @@
+    """
+    # TODO: remove this once the relationships are in the DB, and then automate
+    return [
+        ["task_states", "*--?", "task_events"],
```
Even if these are temporary, using two constants like ``ZERO_TO_N_REL = *--?`` etc. might be more comprehensible for development & ensures consistency.

kinow replied:
Good idea. Will also use the table name constants, e.g. `CylcSuiteDAO.TABLE_TASK_STATES` :+1:

kinow commented:
Thanks for the review @sadielbartholomew ! Will go through the feedback slowly over the next days in-between working in issues in Cylc projects :)

oliver-sanders commented:
What's the status of this one?

kinow commented:
@oliver-sanders I need to address @sadielbartholomew 's feedback and update the PR. But with my bad memory I had already forgotten about this one, will update in the next days. Thanks!

kinow commented on bin/database-diagram.py:
```diff
@@ -0,0 +1,236 @@
+        # eralchemy needs a file that ends with .er to parse markdown
+        with tempfile.NamedTemporaryFile(suffix=".er") as tf_md:
+            markdown = "\n".join(schema)
+            logger.info(f"Markdown generated: ")
```
>unless calling separately is for ease of placing these on separate lines or something

That's the reason they were put in separate lines. At one point I realized it was useful to have that in the output to copy part of the output and execute in `dot` to look at the generated value (or modify it and update the code).

But having to find the `Final dot file: <...` to copy was annoying.

sadielbartholomew replied:
That makes sense, thanks.

kinow commented on bin/database-diagram.py:
```diff
@@ -0,0 +1,236 @@
+        return cursor.fetchall()
+
+
+def is_orphan(table_name, relationships):
```
I immediately thought it was a good idea, but in the end this was the best I could do?

```python
diff --git a/bin/database-diagram.py b/bin/database-diagram.py
index 3ed2e79..a561a8b 100755
--- a/bin/database-diagram.py
+++ b/bin/database-diagram.py
@@ -96,13 +96,6 @@ def get_columns_metadata(table_name, conn):
         return cursor.fetchall()
 
 
-def is_orphan(table_name, relationships):
-    for relationship in relationships:
-        if table_name in relationship:
-            return False
-    return True
-
-
 def schema_to_markdown(db_name):
     """Return the database markdown schema.
 
@@ -115,6 +108,10 @@ def schema_to_markdown(db_name):
     lines = []
     orphans = []
     relationships = get_relationships()
+    tables_related = [
+        table_name
+        for table_tuple in relationships
+        for table_name in [table_tuple[0], table_tuple[2]]]
     with closing(sqlite3.connect(db_name)) as conn:
         with closing(conn.cursor()) as cursor:
             cursor.execute("SELECT name FROM sqlite_master WHERE type='table'")
@@ -127,7 +124,7 @@ def schema_to_markdown(db_name):
                     line = f"\t{pk}{column[1]} {label}"
                     lines.append(line)
 
-                if is_orphan(table, relationships):
+                if table not in tables_related:
                     orphans.append(table)
             # in the eralchemy example, relationships go at the end of the file
             lines.extend([" ".join(value) for value in relationships])
```

Was there another way that would eliminate 115-116?

oliver-sanders replied:
```python
if not any((
    table_name in relationship
    for relationship in relationships
))
```

kinow replied:
Worked like a charm! Out of curiosity, if I had 50 relationships, and 100 tables... wouldn't it be better to store the list of "tables related" instead as before?

Though this one looks more Python-like, I feel like I am iterating the relationships more? :confused: or maybe it's my brain telling me it's past midnight.

kinow replied:
(branch updated anyway :+1: )

oliver-sanders replied:
> wouldn't it be better to store the list of "tables related" instead as before?

If you changed from using a list to using a set, in theory yes (*that is assuming that Python set lookups are actually O(1)*).

Building the list beforehand gives us:

* iterate O(N)
* FOR O(M)
  * lookup O(1)

= O(N + M)

Using the `any` approach give us:

* FOR O(M)
  * iterate O(N)
      * lookup O(1)

= O(MN)

kinow commented:
>I appreciate all your efforts to make the overall dimensions more appropriate, but now the schematic seems to be too wide in that the text comes out very small, so that most would likely struggle to read it, & even opening it from the docs into a separate tab, it needs zooming in to be legible.

This seems to be a bit tricky now @sadielbartholomew , especially as I don't have a lot of experience with dot files.

I found that if I created a subgraph, that included the `broadcast_states` table, then Graphviz aligned the two graphs in the manner you've seen above.

I tried a naive approach of adding `task_jobs` instead to the subgraph, but it got all messed up.

![cylc-database](https://user-images.githubusercontent.com/304786/64027565-0ec70380-cb95-11e9-85c2-d2e57d367c2f.png)

I think I will need to read some more about graphviz/dot. Will update it in the next days :+1:

oliver-sanders commented:
A simple solution might be to transpose the graph.

Try changing the [rankdir](https://graphviz.org/doc/info/attrs.html#d:rankdir) (BTW that page is the Graphviz gosbel).

kinow commented:
Having some pointers like this is going to be extremely helpful, thanks @oliver-sanders !

sadielbartholomew commented:
> This seems to be a bit tricky now @sadielbartholomew , especially as I don't have a lot of experience with dot files.

Ah I see, no problem @kinow; if it is generally really quite difficult to change the layout as per my review comments, then I am happy we can put this in as-is, though perhaps you can generate the overall image very large so that readers of the docs can at least comprehend everything when they right click & hit 'Open image in a new tab' (or other-browser equivalent), & can zoom in to full resolution.

Saying that, it might be possible to fix via Oliver's suggestions above. I propose you give those a try & report back whether you can get anywhere towards addressing the problem without undue time or effort. If not, we can merge & perhaps put up an Issue to convey that eventually we want to make the  resulting image width-to-height ratio closer to one for readability?

kinow commented:
>Saying that, it might be possible to fix via Oliver's suggestions above. I propose you give those a try & report back whether you can get anywhere towards addressing the problem without undue time or effort. If not, we can merge & perhaps put up an Issue to convey that eventually we want to make the resulting image width-to-height ratio closer to one for readability?

Very sensible approach, +1 @sadielbartholomew . Hopefully I will learn some more and update the PR in the next day or so with a possible fix. Thanks!!!

kinow commented:
Oh, had some spare time during breakfast to read about ranks, rank direction, subgraphs, etc. Interesting (and very hard!). 

Here's an updated diagram.

![cylc-database](https://user-images.githubusercontent.com/304786/64055190-17442c00-cbdf-11e9-86d2-23691aaaccba.png)

But the only solution I found was - after reading few examples with it - to use an invisible edge. I thought it was hack-ish, but looking at these examples, it seems to me that it is common practice?

PR updated with the last change.

sadielbartholomew commented:
Excellent, great job @kinow. An invisible edge solution seems fine (& even ideal) by me, since I also believe it is common practice; from my investigations & learning on Graphviz, the (e.g. dot) node-edge layout engine will only on its own produce one state, & then there is a small set of varied final states based on tweaking layout settings, so if you want something specific for nodes & edge arrangement outside of those you need to force the layout engine to include those requirements, e.g. with invisible edges. Although it seems "hacky" on the face of it, in context I do not think it is at all.

Is this ready for re-review now, or do you plan to make any more changes (in case I would be going in to review prematurely)?

kinow commented:
Thanks for asking @sadielbartholomew , I was pondering about the part about giving users some sort of help for reading the diagrams... I feel like we shouldn't document everything about how to read this database diagram, but I agree some hints would be useful.

If we really need to document how to read the diagram, then would it be enough to link the [Wiki for ERD](https://en.wikipedia.org/wiki/Entity%E2%80%93relationship_model), and point that our diagram's notation is close to the min, max variation? (which is funny, as the tool has ER in the name for Erwin, but the diagrams produced in Erwin are different, and not even listed on that Wikipedia :confused:

oliver-sanders commented:
Hi, I've had a go at wrapping this script in a Sphinx extension so it builds automatically.

https://github.com/kinow/cylc-doc/pull/1

kinow commented:
>Hi, I've had a go at wrapping this script in a Sphinx extension so it builds automatically.

Thanks @oliver-sanders ! I hadn't thought about a Sphinx extension. And the code got a lot better without the files manipulation :clap: 

I had to remove the link as the generated file had no link. I suspect the image being the only thing after the anchor and also being a literal block, Sphinx ignored it.

The dependency to `realchemy` was moved to `install_requires` too, as otherwise `make html` would fail (it was an optional in the `[database]` group or `[all]` for pip).

Also added a link to the Wikipedia and changed the text a bit. Here's what it looks like now:

![image](https://user-images.githubusercontent.com/304786/64137102-398db200-ce4a-11e9-95e5-78fc6073f11c.png)

oliver-sanders commented on src/ext/database_diagram.py:
```diff
@@ -0,0 +1,188 @@
+    Returns:
+        list: with relationships in the eralchemy markdown syntax.
+    """
+    # TODO: remove this once the relationships are in the DB, and then automate
```
Is there an issue on Cylc Flow for this?

kinow replied:
I tried to add the relationships there, but doing that I had no idea whether it would break for users with existing databases.

I think it should work just fine, though it could slow down things in case SQLite starts doing some constraint checking maybe. The lowest risk option was to avoid touching the schema, and instead just keep the relationships in this script :disappointed_relieved:

oliver-sanders replied:
I would have though existing databases would be OK, else we could add an upgrader. Speed impact I'm not too sure about.

@matthewrmshin any reason we haven't added foreign keys into the DB yet?

matthewrmshin replied:
Happy to add foreign keys.

IIRC, @kinow already tried that, but there were too many test battery failures that were taking too much effort to resolve.

kinow replied:
Oh yeah, there were the functional tests failing too. That was a bit scary, even though I think it failed because it was comparing the output of the database structure. I can create a ticket for this TODO so that this doesn't slip, and we can try again later? At least once that's done, we have all the tooling here, and it should be easier to implement (and to undo in case something goes south)
